app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,242 @@
+            val message = constructMessage(mServiceHandler?.obtainMessage(), MESSAGE_AVAILABLE_NETWORK)
+            mServiceHandler?.sendMessage(message)
+        }
+    }
+
+    fun getNetworkList(): MutableSet<Network> {
+        return currentNetworkList
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int): Message {
+        val message = Message()
+        message.what = what
+        if (what == MESSAGE_ACTIVE_NETWORK) {
+            message.obj = connectivityManager.activeNetwork
+        }
+        return message
+    }
+
+    private fun constructMessage(message: Message?, what: Int): Message {
+        var messageValue = message
+        if (messageValue == null) {
+            messageValue = Message()
+        }
+        messageValue.what = what
+        if (what == MESSAGE_ACTIVE_NETWORK) {
+            messageValue.obj = connectivityManager.activeNetwork
+        }
+        return messageValue
+    }
+
+    // Handles the network messages from the callback from the connectivity manager
+    private class NetworkRequestHandler(looper: Looper?, val connectionMonitor: ConnectionMonitor) : Handler(looper!!) {
+        var addedNetworkList: LinkedHashSet<Network> = linkedSetOf()
+        override fun handleMessage(msg: Message) {
+            super.handleMessage(msg)
+            processMessage(msg)
+        }
+
+        fun processMessage(msg: Message) {
+            if (msg.what == MESSAGE_ACTIVE_NETWORK) {
+                processActiveNetwork(msg)
+            } else if (msg.what == MESSAGE_AVAILABLE_NETWORK) {
+                processAllAvailableNetwork()
+            }
+        }
+
+        /**
+         * Handles the active networks.
+         * Message object will have the active network associated with it.
+         * Checks if the currentActiveNetwork object is not equal to activeNetwork - VPN underlying
+         * network will be set.
+         */
+        private fun processActiveNetwork(msg : Message){
+            var activeNetwork: Network? = null
+            if (msg.obj != null) {
+                activeNetwork = msg.obj as Network
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message ${connectionMonitor.connectivityManager.getNetworkInfo(activeNetwork)?.typeName.toString()}")
+                if (connectionMonitor.currentActiveNetwork?.networkHandle != activeNetwork.networkHandle) {
+                    connectionMonitor.networkListener.onUpdateActiveNetwork(activeNetwork)
+                    connectionMonitor.currentActiveNetwork = activeNetwork
+                }
+            } else {
+                connectionMonitor.networkListener.onNetworkDisconnected()
+                connectionMonitor.currentActiveNetwork = null
+            }
```

Comment:
Updated.

Reply:
What was updated?

> This can be simplified, there's no need to a separate listener.onUpdateActiveNetwork

I still see `NetworkListener#onUpdateActiveNetwork` being used in the new review diff.

Reply:
Removed the onUpdateActiveNetwork(). Instead calling onNetworkConnected()